File: nodejs-version/lib/utils/signature.js
import * as crypto from 'crypto';

// step 1
function filter_params(queries) {
    return Object.keys(queries).sort().map(function(k) {
        return encodeURIComponent(k) + "=" + encodeURIComponent(queries[k])
    }).join("&")
}

// step 2
function get_str_to_sign(http_verb, path, filtered_params) {
    return http_verb + "\n" + path + "\n" + filtered_params;
}

// step 3
function derive_key(timestamp, appsecret) {
    return crypto.createHmac('sha256', timestamp + "").update(appsecret).digest('hex');
}

export class Signature {
    constructor(appid, appsecret) {
        this.appid = appid;
        this.appsecret = appsecret;
    }

    sign(http_verb, path, params, timestamp) {
        let filtered_params = filter_params(params);
        let str_to_sign = get_str_to_sign(http_verb, path, filtered_params);
        let derived_key = derive_key(timestamp, this.appsecret);

        return crypto.createHmac('sha256', derived_key).update(str_to_sign).digest('hex');
    }


}

File: ruby-version/ppninja/lib/ppninja/signature.rb
require 'openssl'

module Ppninja
  class Signature
    attr_reader :appsecret
    def initialize(appsecret)
      @appsecret = appsecret
    end

    def sign(queries, http_method, path, timestamp)
      # 1. filter parameters
      queries = {} unless queries.is_a?(Hash)
      parameters = ppj_normalization_parameters(queries)
      # 2. get string_to_sign
      string_to_sign = http_method + "\n" + path + "\n" + parameters

      encrypt(string_to_sign, timestamp)
    end

    def encrypt(string_to_sign, timestamp)
      # 1. get derived_key
      digest = OpenSSL::Digest.new('sha256')
      derived_key = OpenSSL::HMAC.hexdigest(digest, timestamp.to_s, @appsecret)

      # 2. using derived_key to sign the string_to_sign
      signature = OpenSSL::HMAC.hexdigest(digest, derived_key, string_to_sign)

      # return
      signature
    end

    private

    def ppj_normalization_parameters(hash)
      queries = hash.keys.sort.map do |k|
        next if /^x-ppj-*/.match(k.downcase) # do not encode parameters used by encrption
        URI.encode(k.to_s) + '=' + URI.encode(hash[k].to_s)
      end
      (queries - [nil, '']).join('&')
    end
  end
end
